ARC-0018 ASA Transfer Royalty Enforcement Specification (#70)
algorandfoundation/ARCs: ARCs/arc-0018.md

A Royalty Policy is specified as a Royalty Collector and associated Royalty weighting. Additionally, if any assets besides Algos are acceptable, 

```ts
type RoyaltyPolicy {
    share:              number,  // The share of the payment due, specified in basis points
    royalty_collector:  string,  // The address that should collect this share of the payment
    allowed_assets:     number[] // The list of asset ids that are acceptable for payment
}
```

### Royalty Enforcer

The Royalty Enforcer is an application that controls the transfer of ASAs subject to the Royalty Policy. 

This is accomplished by exposing an interface defined as a set of [ABI Methods](#abi-methods) allowing a grouped transaction call.

### Royalty Collector

The Royalty Collector is a generic account that could be set to a Single Signature, a Multi Signature, a Smart Signature or even to another Smart Contract. The Royalty Collector is then responsible for the royalty distribution logic, making the Royalty Enforcement Specification more general and composable.

### Royalty Asset 

The Royalty Asset is an ASA subject to royalty collection and MUST be created with the the appropriate roles and parameters. 

> > Because the protocol does not allow updating a role after it's been deleted, if the creator thinks she may want to modify them later, they must be set to some non-zero address.

Reviewer: Are we talking explicitly about the clawback address here?

Author: clawback/freeze both I think

created: 2022-02-16
requires: ARC-4
---

## Abstract

A specification to describe a mechanism to enforce [Royalty Payments](https://en.wikipedia.org/wiki/Royalty_payment) to a Royalty Collector given a policy describing the royalty shares, both on primary and secondary sales.

## Motivation

This ARC is defined to provide a consistent set of asset configurations and ABI methods that, together, enable a royalty payment to a Royalty Collector.

An example may include some music rights where the label, the artist, and any investors have some assigned royalty percentage that should be enforced on transfer. During the sale transaction, the appropriate royalty payments should be included or the transaction must be rejected.

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

[Royalty Policy](#royalty-policy) - The data structure that defines how royalties are dispersed

[Royalty Enforcer](#royalty-enforcer) - The contract that enforces the royalty payments given the Royalty Policy and performs transfers of the assets.

> [Royalty Collector](#royalty-collector) - The account that collect the royalty and provides the royalty distribution logic. It could be set to a Single Signature, a Multi Signature, a Smart Signature or even another Smart Contract.

Reviewer 2: > One thing that we may want to require however is that it necessarily opted in to the allowed assets for payments.
> Also, we may want to specify what happens if it does not:
> does it mean the asset is no more transferrable? or that royalties are not collected?

[From my comment here](https://github.com/algorandfoundation/ARCs/pull/70#discussion_r838331343): I want to avoid (and be able to handle somehow) the "lock-in effect" by which people are no longer able to transfer and trade their ASA not because they refuse to comply with the `Royalty Policy` but because that Policy is broken by the creator ex-post.

If `Royalty Collector` opts-out all the `Share ASA` I would expect that the `Royalty ASA` is now “royalty free” (since the `Royalty Collector` doesn't want to receive Share ASA at all).

[Asset Subject to Royalties](#royalty-asset) - The asset that is subject to royalty payments on transfer

[Third Party Marketplace](#third-party-marketplace) - A third party marketplace may be any marketplace that implements the appropriate methods to initiate transfers.

### Royalty Policy 

A Royalty Share consists of an `address` that should receive a Royalty payment and a `share` representing some share of the total payment amount. The Royalty Share MUST be specified in terms of [basis points](https://en.wikipedia.org/wiki/Basis_point) of the payment amount.

A Royalty Policy is specified as a Royalty Collector and associated Royalty weighting. Additionally, if any assets besides Algos are acceptable, 

```ts
type RoyaltyPolicy {

>     share:              number,  // The share of the payment due, specified in basis points

Reviewer 2: Yes, I think we shouldn't specify a range for this at all. Maybe is not common but what if I want to set a royalty of 150%? I think we should stay general and not restrict this. Each implementation then must just make sure that the calculation is correct (does not overflow, etc.).

>     share:              number,  // The share of the payment due, specified in basis points

Author: Where would the extra 50% payment come from?

>     share:              number,  // The share of the payment due, specified in basis points

Reviewer 2: The situation I'm describing here is the following: suppose for a given Royalty ASA I want to enforce a "fixed" royalty (or tax), regardless the amount of specific trades. Example: the price of the trade is 100 _units_ but the fixed "tax" on transfer is always 150 _units_.

I know it sounds uncommon, but should this spec accomodate this use-case too?

>     share:              number,  // The share of the payment due, specified in basis points

Author: I don't think that'd be a common enough case to address in this spec

    royalty_collector:  string,  // The address that should collect this share of the payment

>     allowed_assets:     number[] // The list of asset ids that are acceptable for payment

Author: Algo would not need to be specified, by default its accepted. This list is just for ASA ids that may also be accepted

>     allowed_assets:     number[] // The list of asset ids that are acceptable for payment

Reviewer: Is it possible to only accept an ASA and not Algos?

>     allowed_assets:     number[] // The list of asset ids that are acceptable for payment

Author: Not as written, but I guess we could add it? I like the idea that there will always be one possible payment that will always be accepted though.

}
```

### Royalty Enforcer

The Royalty Enforcer is an application that controls the transfer of ASAs subject to the Royalty Policy. 

This is accomplished by exposing an interface defined as a set of [ABI Methods](#abi-methods) allowing a grouped transaction call.

### Royalty Collector

The Royalty Collector is a generic account that could be set to a Single Signature, a Multi Signature, a Smart Signature or even to another Smart Contract. The Royalty Collector is then responsible for the royalty distribution logic, making the Royalty Enforcement Specification more general and composable.

### Royalty Asset 

The Royalty Asset is an ASA subject to royalty collection and MUST be created with the the appropriate roles and parameters. 

> Because the protocol does not allow updating a role after it's been deleted, if the creator thinks she may want to modify them later, they must be set to some non-zero address.

#### Asset Roles

> The Clawback role MUST be set to the Application Address.

Reviewer 2: As you say, this introduce complexity but maybe we can be permissive, like:

```
The Clawback role **MUST** be set to the Application Address of the Royalty Enforcer or to any account rekeyed to the Royalty Enforcer.

The first option **SHOULD** be always preferred when possible.
```

I'm wondering if this complicates or not the "discovery" phase for wallets/app that want to learn how to interact with the Royalty ASA.

The Freeze role MUST be set to the Application Address if `FreezeAddr != ZeroAddress`, else set to `ZeroAddress`.

> If the creator wants to allow an ASA to be Royalty Free after some conditions are met, it should be set to the Application Address 

The Manager role MUST be set to the Application Address if `ManagerAddr != ZeroAddress`, else set to `ZeroAddress`.

> If the creator wants to update the Freeze role, this should be set to the application address

The Reserve role MAY be set to anything.

#### Asset Parameters

The `DefaultFrozen` MUST be set to true.

### Third Party Marketplace

In order to support secondary sales on external markets this spec was designed such that the Royalty Asset may be listed without transferring it from the current owners account.  A Marketplace may call the transfer request as long as the address initiating the transfer has been set as the `authorized_address` through the [offer](#offer) method in some previous transaction by the current owner.

### ABI Methods

#### Required

##### SetPolicy
    

>     set_policy(royalty_asset: asset, royalty_receiver: address, royalty_share: uint64, allowed_asset_1: asset, allowed_asset_2: asset, allowed_asset_3: asset, allowed_asset_4: asset)

Author: I think we'd want to pass it in this case since we may want to check its parameters and opt the application into the asset.

Sets the Royalty Policy for a given asset. 

> The `royalty_asset` is the ASA id of the asset the policy is set for.  If the `royalty_asset` is 0, the policy is defined for _any_ asset that has its freeze and clawback set to the application address. 

Reviewer 2: > the policy is defined for _any_ asset that has its freeze and clawback set to the application address.

We should not mandate `FreezeAddress` AND `ClawbackAddress` set to Application Address, but just `ClawbackAddress` since the `FreezeAddress` could be set to `ZeroAddress`.

> The `royalty_asset` is the ASA id of the asset the policy is set for.  If the `royalty_asset` is 0, the policy is defined for _any_ asset that has its freeze and clawback set to the application address. 

Author: @fabrice102 this would be default if no specific policy is defined for a given Asset ID

The `royalty_share` is the share of the payment that should be sent to the `royalty_receiver`.  It is specified in basis points (e.g., 1% is 100).  

The `allowed_assets_n` arguments represents the ASA ids that are acceptable for payment. The contract logic MUST opt into the set of assets specified in order to accept the payments as part of a transfer.

> TODO: since assets cant be in dynamic arrays, this bit is annoying. We could define N selectors (up to 8 for assets) and dispatch them all to the same set of logic? for now just hardcoded to 4 and caller expected to pass 0s if they don't need all 4.  As of https://github.com/algorandfoundation/ARCs/pull/77, this limitation has been changed. Update spec accordingly once SDK support is available.

>The Royalty Collector SHOULD be opted into the full set of assets contained in this list but MAY NOT be checked ahead of time. In the case that an account is not opted into an asset, any transfers where payment is specified for that asset will fail until the account opts into the asset.

##### Transfer: 

>     transfer(royalty_asset: asset, from: account, to: account, royalty_collector: account, royalty_asset_amount: uint64, payment: transaction)

Reviewer 3: What happens of not all the offered amount is used?
Can a second transfer take place?
Or does it rescind the offer?

Author: In the current reference implementation the offered amount is updated to reflect the new offered amount but maybe we should be more explicit about it? It may also make sense to just wipe it completely.

I'm not sure what the correct behavior is, wdyt?

Author: I've updated the document to describe how the implementer might consider updating the offer after a transfer takes place.

Transfers the Asset after checking that the royalty policy is adhered to. This call must be sent by the current owner (the `from` parameter)

The `royalty_asset` is the ASA ID to be transferred. It MUST have a policy defined prior to attempting a transfer.

The `from` parameter is the address for the account the ASA is transferred from.

The `to` parameter is the address for the account the ASA is transferred to.

The `royalty_receiver` parameter is the address for the account that collects the royalty and takes care of any redistribution logic (if any). 

The `royalty_asset_amount` parameter is the number of units of this ASA ID to transfer. The amount MUST be less than or equal to the amount `offered` by the `from` account.

The `payment` parameter is a reference to the transaction that is transferring some asset (ASA or Algos) from the buyer to the Application Address. 

The `payment_asset` parameter is used in the case that the payment is being made with some asset rather than with Algos. It MUST match the Asset ID of the AssetTransfer payment transaction. 

The transfer call SHOULD be part of a group with a size of 2 (app call + payment/asset transfer)

> > While its possible to enforce the group size limit, it is possible to circumvent the royalty enforcement logic by simply making an Inner Transaction application call with the appropriate parameters and a small payment, then in the same outer group the "real" payment. The counter-party risk remains the same since the inner transaction is atomic with the outers. 

Author: For the lower limit we'd have to have it specified for each Allowed Asset defined in the policy.  If some volatile asset is allowed the lower limit may be outside the perceived value of the Royalty Asset being transferred. 

In general I think I'd leave this to the implementer to decide _if_ a transfer should take place given some additional custom logic.

> Some ways to prevent a small royalty payment and larger payment in a later transaction of the same group might be by using an `allow` list that is checked against the `auth_addr` of the offer call. The `allow` list would be comprised of known and trusted marketplaces that do not attempt to circumvent the royalty policy. The `allow` list may be implicit as well by transferring a specific asset to the `auth_addr` as frozen and on `offer` a the balance must be > 0 to allow the `auth_addr` to be persisted.

> Upon each transfer the the Royalty Enforcer SHOULD assert that `AuthAddr == AssetSender` for the Seller (`from` parameter) and `AuthAddr == AssetReceiver` for the Buyer (`to` parameter). In other words Seller and Buyer accounts not have been Rekeyed. Exceptions can be handled with `Move` method.

Reviewer 4: I'm not clear exactly what the intention is here. Do you want to ensure that the Seller or Buyer has never been rekeyed? Or just not within the current group?

If it is a requirement that the accounts have never been required this seems unreasonable given that Rekeying is a very commonly used standard feature of Algorand.

Author: Right, this a non-required bit of the spec that calls out a possible way to effectively transfer the asset without going through whatever pre-requisites are required by the contract. 

A holder of this asset (or more generally any controlled asset, think of a security) might need to go through some KYC/AML that allows them to hold it. Using the rekey feature in an atomic transaction allows an account to atomically "sell" the asset.

I don't want to make it required here and think most implementers probably don't need to do this

> > The rational for this check is ensuring that `Rekeying` is not exploited to transfer the `royalty_asset` ownership circumventing the Royalty Policy. Two parties involved in the trade could create a Group Transactions in which: 1. Seller rekeys the `from` to the Buyer `to`, 2. Buyer sends payment to `other_seller_account`. Note that nothing can prevent the users to execute such trade but the Royalty Enforcer tries to dissuade this behavior: if such trade happens, the Receiver will no longer be able to re-sell the Royalty ASA in the regular way.

Reviewer 3: This does not prevent the use of smart contract controlling the account holding the asset.
That being said, such an attack requires all the users to agree on using such a smart contract from the beginning, so I guess this is unlikely to be a real threat.

#### Offer:

>     offer(royalty_asset: asset, royalty_asset_amount: uint64, authorized_address: account)

Reviewer 3: What happens if called multiple times?

Flags the asset as transferrable and sets the address that may initiate the transfer request.

The `royalty_asset` is the ASA ID that is offered. It MUST have a policy defined prior to attempting to offer

The `royalty_asset_amount` is the number of units of the ASA ID that are offered. The account making this call MUST have at least this amount. 

> The `authorized_address` is the address that may initiate the transfer. This address may be any valid address in the Algorand network including and Application Account's address.

Reviewer 3: ```suggestion
The `authorized_address` is the address that may initiate the transfer. This address may be any valid address in the Algorand network including an Application Account's address.
```

If any transfer is initiated by an address that is _not_ listed as the `authorized_address` for this asset ID from this account, the transfer is rejected.

#### Rescind:

    rescind(royalty_asset: asset)

Removes the asset from being offered by the sender.

The `royalty_asset` is the ASA ID that is being rescinded.

#### Optional

> #### Create NFT

Author: I put it in the `optional` section but I can remove completely if you think it might be confusing

    create_nft()uint64

Creates an NFT with the appropriate parameters set. 

Returns the Asset ID that was just created

##### Move 

>     move(royalty_asset: asset, royalty_asset_amount: uint64, to: account)

Reviewer 3: Maybe `royalty_free_move`?

>     move(royalty_asset: asset, royalty_asset_amount: uint64, to: account)

Reviewer 4: Is this optional or required? I think the heading levels are confused in this section.

The existence of this method seems to negate any concerns about rekeying or inner txns or allow lists given that there is an explicit method for royalty free transfers. 

If this is optional what is the intended flow with assets without this method for marketplaces that use escrows? Or for users moving between their own accounts? Is a 0 amount transfer allowed and recommended here?

Author: This is optional and acts as an escape valve to transfer the asset without employing Royalty Enforcement logic.

It can be useful for "freeebies" or just to distribute the asset initially. Provided here as an implementation suggestion but not requirement

Moves an asset to the new address without enforcing any royalty payment. May be useful in the case of a marketplace where the NFT must be placed in some escrow account. Any logic may be used to validate this is an authorized transfer.

The `royalty_asset` is the asset being transferred without applying the Royalty Enforcement logic.

The `royalty_asset_amount` is the number of units of this ASA ID that should be moved.

The `to` parameter is the intended receiver of the asset. This account MUST be opted into the asset prior to calling this method or it will fail at the protocol level.

### Storage

> #### Global Storage

Reviewer 3: Specifying global storage is good for two reasons:
1. allowing users to use it to get info (which requires to enter into the level of details you provide)
2. allowing users to know when to include a foreign account in the transaction (which only requires to specify which application call requires which local account).

2 is very important.
1 is useful for actual applications but I'm afraid that in a standard / interface it may prevent some optimizations that applications may do.
Instead, using read-only methods may allow more flexibility.

What do you think?

Author: I agree, I've added read only methods for both policy and offer

                {"name":"allowed_asset_3", "type":"asset"},
                {"name":"allowed_asset_4", "type":"asset"}
            ],
            "returns":{"type":"void"}
        },
        {
            "name":"transfer",
            "desc":"Transfers an Asset from one account to another and enforces royalty payments. If payment is in Algos, payment_asset should be 0",
            "args":[
                {"name":"royalty_asset", "type":"asset"}, 
                {"name":"owner_acct", "type":"account"},
                {"name":"buyer_acct", "type":"account"},
                {"name":"royalty_receiver", "type":"account"},
                {"name":"royalty_asset_amount", "type":"uint64"},
                {"name":"payment", "type":"txn"},
                {"name":"payment_asset", "type":"asset"}
            ],
            "returns":{"type":"void"}
        }
    ],
    "locals": [ 
        {
            "name": "{uint64 asset id as big endian bytes}",
            "desc": "The ID of the asset that has been offered as key, authorizing address as value",
            "type": "(address,uint64)"
        }  
    ],
    "globals": [ 
        {

>             "name": "0000000000000000",

Reviewer 3: Is it really this or the 0 byte to be consistent with the royalty policy for asset ID in key

Reviewer 3: See also https://github.com/algorandfoundation/ARCs/pull/75#discussion_r839006561

Author: In the current implementation, its 8 0 bytes representing the uint64 zero value to be consistent. We could rename to just `default` or something?

            "desc": "The default royalty policy for any ASA unless its ASA ID is specified",
            "type": "(address,uint64,uint64,uint64,uint64)"
        },
        {
            "name": "{uint64 asset id as big endian bytes}",
            "desc": "Royalty policy for asset ID in key ",
            "type": "(address,uint64,uint64,uint64,uint64)"
        }  
    ]
}
```

### Metadata

The metadata associated to an asset MAY conform to [ARC-3](./arc-0003.md) or [ARC-69](./arc-0069.md) with an additional field in the `attributes` section specifying the Application ID for the Royalty Enforcer Contract. 

```js
//...

> "attributes":{

Reviewer 3: I think we should use `properties`, not `attributes`.
That would also be better re https://github.com/algorandfoundation/ARCs/pull/61/files

One issue is that ARC-69 only allows simple properties, but maybe we can convince them otherwise.

```

## Rationale

The motivation behind defining a Royalty Enforcement specification is the need to guarantee a portion of a payment is received by select royalty collector on sale of an asset. Current royalty implementations are either platform specific or are only adhered to when an honest seller complies with it, allowing for the exchange of an asset without necessarily paying the royalties.

The use of a smart contract as a clawback address is a guaranteed way to know an asset transfer is only ever made when certain conditions are met, or made in conjunction with additional transactions. The Royalty Enforcer is responsible for the calculations required in dividing up and dispensing the payments to the respective parties. The present specification does not impose any restriction on the Royalty Collector distribution logic (if any), which could be achieved through a Multi Signature account, a Smart Signature or even through another Smart Contract.

On Ethereum the EIP-2981 standard allows for ERC-721 and ERC-1155 interfaces to signal a royalty amount to be paid, however this is not enforced and requires marketplaces to implement and adhere to it.

## Backwards Compatibility

> Existing ASAs with unset freeze and clawback roles will be incompatible with this specification.

Reviewer 3: ```suggestion
Existing ASAs with unset clawback address or unset manager address (in case the clawback address is not the application account of a smart contract that is updatable - which is most likely the case) will be incompatible with this specification.
```

## Reference Implementation

https://github.com/algorand-devrel/royalty

> ## Security Considerations

Reviewer 3: See also comments above.

### Royalty Asset 

The Royalty Asset is an ASA subject to royalty collection and **MUST** be created with the the appropriate mutable and immutable parameters. 

> Because the protocol does not allow updating an address parameter after it's been deleted, if the creator thinks they may want to modify them later, they must be set to some non-zero address.

#### Asset Address Parameters 

The Clawback parameter **MUST** be set to the Application Address of the Royalty Enforcer.

> Since the Royalty Enforcer relies on using the Clawback mechanism to perform the transfer the Clawback should NEVER be set to the zero address. 

The Freeze parameter **MUST** be set to the Application Address of the Royalty Enforcer if `FreezeAddr != ZeroAddress`, else set to `ZeroAddress`.

> If the creator wants to allow an ASA to be Royalty Free after some conditions are met, it should be set to the Application Address 

The Manager parameter **MUST** be set to the Application Address of the Royalty Enforcer if `ManagerAddr != ZeroAddress`, else set to `ZeroAddress`.

> > If the creator wants to update the Freeze parameter, this should be set to the application address

Reviewer 5: What happens if a marketplace needs to burn an NFT? If the manager address must be the Application Address, then the Royalty Enforcer needs to be able to burn the NFT as well, or I guess that could be an optional feature.

Reviewer 2: Yes this standard does not mandate what you could do as Manager through the Application, this is a free choice of the implementation (e.g. could let Royalty ASA "admins" act like a Manager using the Application as a proxy).

### ABI Methods

#### Required

##### Set Policy
    
```
set_policy(royalty_asset: asset, royalty_receiver: address, royalty_share: uint64, allowed_asset_1: asset, allowed_asset_2: asset, allowed_asset_3: asset, allowed_asset_4: asset)
```

Sets the Royalty Policy for a given asset. 

The `royalty_asset` is the ASA id of the asset the policy is set for.  If the `royalty_asset` is 0, the policy is defined for _any_ asset that has its clawback set to the application address. 

The `royalty_receiver` is the address of the account that should receive a partial share of the payment or asset transfer for any [transfer](#transfer) of the `royalty_asset` 

The `royalty_share` is the share of the payment that should be sent to the `royalty_receiver`.  It is specified in basis points (e.g., 1% is 100).  

The `allowed_assets_n` arguments represents the ASA ids that are acceptable for payment. The contract logic **MUST** opt into the set of assets specified in order to accept the payments as part of a transfer.

> The Royalty Collector **SHOULD** be opted into the full set of assets contained in this list. 

Author: ```suggestion
The Royalty Collector **MUST** be opted into the full set of assets contained in this list. 
```

The `from` parameter is the address for the account the ASA is transferred from.

The `to` parameter is the address for the account the ASA is transferred to.

The `royalty_receiver` parameter is the address for the account that collects the royalty and takes care of any redistribution logic (if any). 

The `royalty_asset_amount` parameter is the number of units of this ASA ID to transfer. The amount **MUST** be less than or equal to the amount `offered` by the `from` account.

The `payment` parameter is a reference to the transaction that is transferring some asset (ASA or Algos) from the buyer to the Application Address of the Royalty Enforcer. 

The `payment_asset` parameter is used in the case that the payment is being made with some asset rather than with Algos. It **MUST** match the Asset ID of the AssetTransfer payment transaction. 

The `curr_offer_amount` parameter is the current amount of the Royalty Asset offered by the `from` account. 

The transfer call **SHOULD** be part of a group with a size of 2 (payment/asset transfer + app call)

See [Security Considerations](#security-considerations) for details on how this method may be gamed.

> Upon each transfer the the Royalty Enforcer SHOULD assert that `AuthAddr == AssetSender` for the Seller (`from` parameter) and `AuthAddr == AssetReceiver` for the Buyer (`to` parameter). In other words Seller and Buyer accounts not have been Rekeyed. Exceptions can be handled with `Move` method.

Author: ```suggestion
Upon each transfer the the Royalty Enforcer **SHOULD** assert that `AuthAddr == AssetSender` for the Seller (`from` parameter) and `AuthAddr == AssetReceiver` for the Buyer (`to` parameter). In other words Seller and Buyer accounts not have been Rekeyed. Exceptions can be handled with `Move` method.
```

### Royalty Receiver

The Royalty Receiver is a generic account that could be set to a Single Signature, a Multi Signature, a Smart Signature or even to another Smart Contract. 

The Royalty Receiver is then responsible for any further royalty distribution logic, making the Royalty Enforcement Specification more general and composable.

### Royalty Asset 

The Royalty Asset is an ASA subject to royalty payment collection and **MUST** be created with the the [appropriate parameters](#royalty-asset-parameters). 

> Because the protocol does not allow updating an address parameter after it's been deleted, 
> if the creator thinks they may want to modify them later, they must be set to some non-zero address.

#### Asset Offer

The Asset Offer is the a data structure stored in the owners local state. It is keyed in local storage by the byte string representing the ASA Id. 

```ts
interface AssetOffer {
    auth_address:   string // The address of a marketplace or account that may issue a transfer request
    offered_amount: number // The number of units being offered
}
```

> This concept is important to this specification because we use the clawback feature to transfer the assets. Without some signal that the current owner is willing to have their assets transferred, it is be possible to transfer the asset without their permission. 

Reviewer 6: it is be possible - > it may be possible

In order for a transfer to occur, this field **MUST** be set and the parameters of the transfer request **MUST** match the value set.

> > A transfer matching the offer would require the transfer amount <= offered amount and that the transfer is sent by auth_address

Reviewer 6: Does this transfer automatically clear the offer afterward?

> If the creator wants to allow an ASA to be Royalty Free after some conditions are met, it should be set to the Application Address 

The Manager parameter **MUST** be set to the Application Address of the Royalty Enforcer if `ManagerAddr != ZeroAddress`, else set to `ZeroAddress`.

> If the creator wants to update the Freeze parameter, this should be set to the application address

The Reserve parameter **MAY** be set to anything.

The `DefaultFrozen` **MUST** be set to true.

### Third Party Marketplace

In order to support secondary sales on external markets this spec was designed such that the Royalty Asset may be listed without transferring it from the current owners account.  

A Marketplace may call the transfer request as long as the address initiating the transfer has been set as the `authorized_address` through the [offer](#offer) method in some previous transaction by the current owner.

### ABI Methods

The following is a set of methods that conform to the [ABI](arc-0004.md) specification meant to enable the configuration of a Royalty Policy and perform transfers.

#### Set Policy  (REQUIRED):
    
```

> set_policy(royalty_basis: uint64, royalty_recipients: account)

Reviewer 6: royalty_recipients to royalty_recipient

```

Sets the policy for any assets using this application as a Royalty Enforcer.

The `royalty_basis` is the percentage for royalty payment collection, specified in basis points (e.g., 1% is 100).  

> A Royalty Basis **MUST** be immutable, if an application call is made that would overwrite an existing value it **MUST** fail. 

Reviewer 6: I thought we were not going to force this and should offer language to put a max in here

> The immutability of a royalty basis is important since mutability introduces the possibility for a situation where, after an initial sale, the royalty policy is updated from 1% to 100% for example. This would make any further sales have the full payment amount sent to the royalty recipient and the seller would receive nothing.

The `royalty_receiver` is the address of the account that should receive a partial share of the payment for any [transfer](#transfer) of an asset subject to royalty collection.

The Royalty Collector **MUST** be opted into the full set of assets contained in this list of payment_assets. 

> In the case that an account is not opted into an asset, any transfers where payment is specified for that asset will fail until the account opts into the asset. or the policy is updated.

> #### Set Payment Asset (REQUIRED):

Reviewer 6: Why are we making this required? If they want to only accept algos

Author: Only required to be implemented, not required to be used

```
set_payment_asset(payment_asset: asset, allowed: boolean)
```

The `payment_asset` argument represents the ASA id that is acceptable for payment. The contract logic **MUST** opt into the asset specified in order to accept them as payment as part of a transfer.

The `allowed` argument is a boolean representing whether or not this asset is allowed.

#### Transfer (REQUIRED): 

```
transfer(
    royalty_asset: asset, 
    from: account, 
    to: account, 
    royalty_receiver: account, 
    royalty_asset_amount: uint64, 
    payment: transaction, 
    payment_asset: asset, 
    current_offer_amount: uint64
)
```

> Transfers the Asset after checking that the royalty policy is adhered to. This call must be sent by the current owner (the `from` parameter)

Reviewer 6: This call must be sent by the current owner - I thought the auth address

The `royalty_asset` is the ASA ID to be transferred. It **MUST** have a policy defined prior to attempting a transfer.

The `from` parameter is the account the ASA is transferred from.

The `to` parameter is the account the ASA is transferred to.

The `royalty_receiver` parameter is the account that collects the royalty payment. 

The `royalty_asset_amount` parameter is the number of units of this ASA ID to transfer. The amount **MUST** be less than or equal to the amount [offered](#offer) by the `from` account.

The `payment` parameter is a reference to the transaction that is transferring some asset (ASA or Algos) from the buyer to the Application Address of the Royalty Enforcer. It **MUST** be one of a Payment transaction or Asset Transfer transaction.

> The `payment_asset` parameter is used in the case that the payment is being made with some asset rather than with Algos. It **MUST** match the Asset ID of the AssetTransfer payment transaction. 

Reviewer 6: And the enforcer must be opted into the asset.

The `current_offer_amount` parameter is the current amount of the Royalty Asset [offered](#offer) by the `from` account. 

> The transfer call **SHOULD** be part of a group with a size of 2 (payment/asset transfer + app call)

Reviewer 6: Are we specifying a specific order

Author: its specified as part of the abi method signature

If this method is called when there is an existing entry for the same `royalty_asset`, the call is treated as an update.  In the case of an update case the contract **MUST** compare the `current_royalty_asset_amount` and `current_authorized_address` with what is currently set.  If the values differ, the call **MUST** be rejected. 

> This requirement is meant to prevent a sort of race condition where the `authorized_address` has a `transfer` accepted before the `offer`-ing account sees the update. In that case the offering account might try to offer more than they would otherwise want to. An example is offered in [security considerations](#security-considerations)

To rescind an offer, this method is called with 0 as the new offered amount.

If a [transfer](#transfer) or [royalty_free_move](#royalty-free-move) is called successfully, the `offer` **SHOULD** be updated or deleted from local state. 

> Exactly how to update the offer is left to the implementer. In the case of a partially filled offer, the amount may be updated to reflect some new amount that represents `offer_amount - amount transferred` or the offer may be deleted completely.

> #### Get Offer (OPTIONAL):

Reviewer 6: I would suggest we remove this

        {
            "name": "royalty_basis",
            "desc": "The percentage for royalty payment specified in basis points",
            "type": "uint64"
        },
        {
            "name": "royalty_recipient",
            "desc": "The receiver of any royalty payment",
            "type": "address"
        }
    ]
}
```

#### Example Flow for a Marketplace

```
Let Alice be the creator of the Royalty Enforcer and Royalty Asset
Let Alice also be the Royalty Recipient
Let Bob be the Royalty Asset holder
Let Carol be a buyer of a Royalty Asset
```

```mermaid
sequenceDiagram
    Alice->>Royalty Enforcer: set_policy with Royalty Basis and Royalty Recipient
    Alice->>Royalty Enforcer: set_allowed_asset with any asset that should be accepted as payment 

    par List
        Bob->>Royalty Enforcer: offer
        Bob->>Marketplace: list
    end

    Par Buy
        Carol->>Marketplace: buy
        Marketplace->>Royalty Enforcer: transfer
        Bob->>Carol: clawback issued by Royalty Enforcer
        Royalty Enforcer->>Alice: royalty payment
    end

    par Delist
        Bob->>Royalty Enforcer: offer 0 
        Bob->>Marketplace: delist
    end
```

### Metadata

> The metadata associated to an asset **MAY** conform to any ARC that supports an additional field in the `properties` section specifying the specific information relevant for off-chain applications like wallets or Marketplace dApps.

Reviewer 6: This should be immutable imo

The fields that should be specified are:

  - `application_id` - The ID of the Royalty Enforcer application

>   - `application_address` - The Address of the Royalty Enforcer application

Reviewer 6: Why do we have these

Reviewer 6: i thought it was just application_id

Author: Just convenience for off chain things to make parsing easier. Otherwise its a couple api calls and parsing out encoded bytestrings.

I can remove but they seem useful?

> This May be useful in the case of a marketplace where the NFT must be placed in some escrow account. Any logic may be used to validate this is an authorized transfer.

The `royalty_asset` is the asset being transferred without applying the Royalty Enforcement logic.

The `royalty_asset_amount` is the number of units of this ASA ID that should be moved.

The `from` parameter is the current owner of the asset.

The `to` parameter is the intended receiver of the asset.

The `current_offered_amount` is the number of units of this asset currently offered. This value **MUST** be greater than or equal to the amount being transferred. 

The `current_auth_address` is the authorized address for the current offer. This value **MUST** be the same as the sender of this transaction (the Royalty Enforcer Administrator).

> The `current_*` values for the offer are passed to prevent the race or attack described in [Security Considerations](#security-considerations).

### Read Only Methods

Three methods are specified here as `read-only` as defined in ARC22.

#### Get Policy
```

> get_policy()(address,uint64)

Reviewer 5: Should this be named to mirror the setter?

```suggestion
get_royalty_policy()(address,uint64)
```

Alternatively, renamed the setter to (the shorter) `set_policy`.

The `royalty_receiver` parameter is the address for the account that collects the royalty and takes care of any redistribution logic (if any). 

The `payment` parameter is a reference to the transaction that is transferring some asset (ASA or Algos) from the buyer to the Application Address. 

Upon each transfer the the Royalty Enforcer MUST assert that `AuthAddr == AssetSender` for the Seller (`from` parameter) and `AuthAddr == AssetReceiver` for the Buyer (`to` parameter). In other words Seller and Buyer accounts not have been Rekeyed. Exceptions can be handled with `Move` method.

> The rational for this check is ensuring that `Rekeying` is not exploited to transfer the `royalty_asset` ownership circumventing the Royalty Policy. Two parties involved in the trade could create a Group Transactions in which: 1. Seller rekeys the `from` to the Buyer `to`, 2. Buyer sends payment to `other_seller_account`. Note that nothing can prevent the users to execute such trade but the Royalty Enforcer tries to dissuade this behavior: if such trade happens, the Receiver will no longer be able to re-sell the Royalty ASA in the regular way.

#### Offer:

>     offer(royalty_asset: asset, authorized_address: account)

Reviewer 2: See this [comment](https://github.com/algorandfoundation/ARCs/pull/70#discussion_r838336925). I think we should add `RoyaltyAssetAmount` here too.
